lua/lualine.lua

```diff
@@ -65,17 +69,54 @@ local function load_special_components(component)
   end
 end
 
+local function component_loader(component)
+  if type(component[1]) == 'function' then return component end
+  -- apply default args
+  for opt_name, opt_val in pairs(M.opt) do
+    if component[opt_name] == nil then
+      component[opt_name] = opt_val
+    end
+  end
+  -- set custom highlights
+  if component.color then
+    local component_name = component[1]
+    local color = component.color
+    local function update_color()
+      component.color = highlight.create_component_highlight_group(color, component_name, component)
+    end
+    update_color()
+    utils.expand_set_theme(update_color)
+  end
+  -- load the component
+  local ok, loaded_component = pcall(require, 'lualine.components.' .. component[1])
+  if not ok then
+    loaded_component = load_special_components(component[1])
+  end
+  component[1] = loaded_component
+  if type(component[1]) == 'table' then
+    component[1] = component[1].init(component)
+  end
+end
+
+
 local function load_components()
   local function load_sections(sections)
-    for _, section in pairs(sections) do
+    for section_name, section in pairs(sections) do
       for index, component in pairs(section) do
-        if type(component) == 'string' then
-          local ok,loaded_component = pcall(require, 'lualine.components.' .. component)
-          if not ok then
-            loaded_component = load_special_components(component)
-          end
-          section[index] = loaded_component
+        if type(component) == 'string' or type(component) == 'function' then
+          component = {component}
         end
+        component.self = {}
+        -- used to provide default for bg and fg for custom highlights
+        component.self.theme = theme_set
+        -- setting highlight because utils need highlight but cann't require
+        -- it as it creates circular dependency . It's a workaround for now
+        -- Should look for a better solution.
+        component.self.highlight = highlight
+        -- convert lualine_section -> section before setting section
+        component.self.section = section_name:sub(9,10)
```

Comment:
I don't really get what's going on here. This comment doesn't say much to me unfortunately

Reply:
I was croping `lualine_a` to `a` to use it as a key to lookup theme table. It doesn't make much sense here . I'll move it to highlight.

Reply:
moved to highlight where it's relevent